### PR TITLE
Stably sort controllerrevisions

### LINKED ISSUE #61998: Statefulset pods sometimes continually rolling
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Statefulsets are sometimes continually rolling pods and I don't know why
statefulset created: 
statefulset.yml:
https://gist.github.com/ryanmcnamara/9e81c2bd2c6ac22fe8c12bc2a68a8aad
storageclass referenced:
https://gist.github.com/ryanmcnamara/cdf8ade20d3a853d5ed2d80b60ead79f
kubectl describe statefulset statefulset-test: 
https://gist.github.com/ryanmcnamara/e2bf74a8d7a4f0cbaa19f98ca57b4348

As you can see from the above describe, the pod in the statefulset has been restarted a lot of times. Observing the created pod, 
1) The pod comes up ready 
2) immediately transitions to terminating 
3) Is terminated
4) A new pod is created
5) Go back to (1) (repeats forever)

Despite no updates to the statefulset spec ocurring.

Making a change to the statefulset template (for example a trivial change adding a nonsense annotation) stops the rolling and fixes the issue

When this statefulset was created, there was a mutating webhook enabled. After the observed pod rolling. I disabled this by making its configuration invalid and making it non required, and the rolling persisted.

The status of the statefulset: `kubectl get statefulset statefulset-test -o yaml`:
https://gist.github.com/ryanmcnamara/1cbd58c4e9b7843f9443ee4010d41ffb

You can see that there is a discrepency between currentRevision and updateRevision.
  currentRevision: statefulset-test-8696d5779b
  updateRevision: statefulset-test-8696d5778f
A short while later the status is changed to contain
  currentRevision: statefulset-test-d8b584c9f
  updateRevision: statefulset-test-d8b584c9b
This change happened at Sun Apr  1 19:48:48 2018
Presumably after the rollout of statefulset-test-8696d5778f is complete.

Here are all the controllerrevisions from above:
kubectl get controllerrevision statefulset-test-8696d5779b -o yaml
https://gist.github.com/ryanmcnamara/ee54d1d5941cfde857a5589a8347c9a5
kubectl get controllerrevision statefulset-test-8696d5778f -o yaml:
https://gist.github.com/ryanmcnamara/4bf24bafa00c8e2434ae5d4287ed728a
kubectl get controllerrevision statefulset-test-d8b584c9b -o yaml:
https://gist.github.com/ryanmcnamara/4b49a0504653ba2d92585fb4015897c2

The differences between the above controllerrevisions are insignificant (the hash, the creation timestamp, the name, the resource version, the uid, the self link)

Interestingly all these controllerrevisions were created very close together in time shortly after I created the statefulset
The current controllerrevisions are
```
[rmcnamara@ip-10-0-112-31 ~]$ kubectl get controllerrevisions
NAME                          CONTROLLER                     REVISION   AGE
statefulset-test-8696d57789   StatefulSet/statefulset-test   1          13h
statefulset-test-8696d5778b   StatefulSet/statefulset-test   1          13h
statefulset-test-8696d5778f   StatefulSet/statefulset-test   1          13h
statefulset-test-8696d5779b   StatefulSet/statefulset-test   1          13h
statefulset-test-cf9d4cdf6    StatefulSet/statefulset-test   1          13h
statefulset-test-cf9d4cdfc    StatefulSet/statefulset-test   1          13h
statefulset-test-cf9d4cdfd    StatefulSet/statefulset-test   1          13h
statefulset-test-cf9d4cf44    StatefulSet/statefulset-test   1          13h
statefulset-test-cf9d4cf45    StatefulSet/statefulset-test   1          13h
statefulset-test-d8b584c9b    StatefulSet/statefulset-test   1          13h
statefulset-test-d8b584c9f    StatefulSet/statefulset-test   1          13h
```
After watching the state change for a while, it actually seems like the statefulset is just cycling through the updateRevision being one of the above 10 controllerrevisions

Leaders during this time
```
[rmcnamara@ip-10-0-112-31 ~]$ kubectl get endpoints kube-controller-manager -n kube-system -o yaml | grep alpha
    control-plane.alpha.kubernetes.io/leader: '{"holderIdentity":"ip-10-0-100-151","leaseDurationSeconds":15,"acquireTime":"2018-04-01T05:50:03Z","renewTime":"2018-04-01T20:13:58Z","leaderTransitions":7}'
[rmcnamara@ip-10-0-112-31 ~]$ kubectl get endpoints kube-scheduler -n kube-system -o yaml | grep alpha
    control-plane.alpha.kubernetes.io/leader: '{"holderIdentity":"ip-10-0-100-151","leaseDurationSeconds":15,"acquireTime":"2018-04-01T05:50:01Z","renewTime":"2018-04-01T20:14:06Z","leaderTransitions":8}'
```

From kube-controller-manager logs during this time, here is one excerpt:
```
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.344768    1565 stateful_set.go:455] Syncing StatefulSet default/statefulset-test with 1 pods
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.344653    1565 disruption.go:403] No PodDisruptionBudgets found for pod statefulset-test-0, PodDisruptionBudget controller will avoid syncing.
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.344861    1565 disruption.go:343] No matching pdb for pod "statefulset-test-0"
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.346127    1565 stateful_set_control.go:502] StatefulSet default/statefulset-test terminating Pod statefulset-test-0 for update
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.353804    1565 disruption.go:340] updatePod called on pod "statefulset-test-0"
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.353766    1565 stateful_set.go:226] Pod statefulset-test-0 updated, objectMeta {Name:statefulset-test-0 GenerateName:statefulset-test- Namespace:default SelfLink:/api/v1/namespaces/default/pods/statefulset-test-0 UID:aa09a5de-35e5-11e8-93f3-0e0de0418f44 ResourceVersion:957880 Generation:0 CreationTimestamp:2018-04-01 19:48:35 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[controller-revision-hash:statefulset-test-cf9d4cdfd deployment-test: statefulset.kubernetes.io/pod-name:statefulset-test-0] Annotations:map[] OwnerReferences:[{APIVersion:apps/v1beta1 Kind:StatefulSet Name:statefulset-test UID:fc4a62e6-3579-11e8-93f3-0e0de0418f44 Controller:0xc423c291f6 BlockOwnerDeletion:0xc423c291f7}] Initializers:nil Finalizers:[] ClusterName:} -> {Name:statefulset-test-0 GenerateName:statefulset-test- Namespace:default SelfLink:/api/v1/namespaces/default/pods/statefulset-test-0 UID:aa09a5de-35e5-11e8-93f3-0e0de0418f44 ResourceVersion:957882 Generation:0 CreationTimestamp:2018-04-01 19:48:35 +0000 UTC DeletionTimestamp:2018-04-01 19:49:16 +0000 UTC DeletionGracePeriodSeconds:0xc42233dde0 Labels:map[controller-revision-hash:statefulset-test-cf9d4cdfd deployment-test: statefulset.kubernetes.io/pod-name:statefulset-test-0] Annotations:map[] OwnerReferences:[{APIVersion:apps/v1beta1 Kind:StatefulSet Name:statefulset-test UID:fc4a62e6-3579-11e8-93f3-0e0de0418f44 Controller:0xc42233dde8 BlockOwnerDeletion:0xc42233dde9}] Initializers:nil Finalizers:[] ClusterName:}.
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.353829    1565 disruption.go:403] No PodDisruptionBudgets found for pod statefulset-test-0, PodDisruptionBudget controller will avoid syncing.
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.353854    1565 disruption.go:343] No matching pdb for pod "statefulset-test-0"
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.354030    1565 event.go:218] Event(v1.ObjectReference{Kind:"StatefulSet", Namespace:"default", Name:"statefulset-test", UID:"fc4a62e6-3579-11e8-93f3-0e0de0418f44", APIVersion:"apps", ResourceVersion:"957843", FieldPath:""}): type: 'Normal' reason: 'SuccessfulDelete' delete Pod statefulset-test-0 in StatefulSet statefulset-test successful
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.354836    1565 attach_detach_controller.go:493] processVolumesInUse for node "ip-10-0-107-133.ec2.internal"
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.354883    1565 actual_state_of_world.go:350] SetVolumeMountedByNode volume kubernetes.io/aws-ebs/aws://us-east-1a/vol-0d5396c605c0a817f to the node "ip-10-0-107-133.ec2.internal" mounted true
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.354900    1565 actual_state_of_world.go:350] SetVolumeMountedByNode volume kubernetes.io/aws-ebs/aws://us-east-1a/vol-079810a7d16ba3e53 to the node "ip-10-0-107-133.ec2.internal" mounted true
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.355413    1565 stateful_set.go:417] Finished syncing statefulset "default/statefulset-test" (10.728123ms)
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: E0401 19:48:46.355436    1565 stateful_set.go:399] Error syncing StatefulSet default/statefulset-test, requeuing: StatefulSet.apps "statefulset-test" is invalid: status.currentReplicas: Invalid value: -1: must be greater than or equal to 0
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.355514    1565 stateful_set.go:455] Syncing StatefulSet default/statefulset-test with 1 pods
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.356747    1565 stateful_set_control.go:351] StatefulSet default/statefulset-test has 1 unhealthy Pods starting with statefulset-test-0
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.356765    1565 stateful_set_control.go:412] StatefulSet default/statefulset-test is waiting for Pod statefulset-test-0 to Terminate
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.362906    1565 stateful_set_control.go:99] StatefulSet default/statefulset-test pod status replicas=1 ready=1 current=0 updated=0
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.362917    1565 stateful_set_control.go:107] StatefulSet default/statefulset-test revisions current=statefulset-test-d8b584c9f update=statefulset-test-8696d5778b
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.362925    1565 stateful_set.go:460] Successfully synced StatefulSet default/statefulset-test successful
Apr 01 19:48:46 ip-10-0-100-151 kube-controller-manager[1565]: I0401 19:48:46.362930    1565 stateful_set.go:417] Finished syncing statefulset "default/statefulset-test" (7.47279ms)
```

The error sync in there is suspicous, and as you can see the controllerrevisions are changing in the logs

I have the full logs from a 5 minute window on either side of the above if necessary.

The reason I'm inclined to think this is a kubernetes bug is that any update to the statefulset template fixes the problem.

**How to reproduce it (as minimally and precisely as possible)**:

Not really sure. Worth just trying to create the above statefulset (requires creating the 
I tried deleting and recreating the above statefulset and it didn't happen, so something flaky is going on

**Anything else we need to know?**:
kubernetes 1.9.3 (client and server)
cloud provider: aws

### PR DESCRIPTION
**Pull request by ryanmcnamara**
Fixes https://github.com/kubernetes/kubernetes/issues/61998

There are times when multiple "equal" controllerrevisions are created with
the same revision number. When this happens and this is the case for the
largest revision number, the statefulset controller will periodically
select one of the maximal controllerrevisions to be the target of the
underlying statefulset. The selection happens here: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L212.
Prior to this change this selection was random as the sort was not
stable, which caused the pods of a stable set to continually roll.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Stably sort controllerrevisions. This can prevent pods of statefulsets from continually rolling.
```

### CODE REVIEW

**Comment by ryanmcnamara** [comment]
This is somewhat related to this bug: https://github.com/kubernetes/kubernetes/pull/63282
Regardless of how the discussion finishes there, this will still cause a problem as there can be multiple equal controller revisions at the end of the sorted array that can be selected as mentioned above, and without the stable sort they are selected randomly

**Comment by ryanmcnamara** [comment]
```release-note
Stably sort controllerrevisions. This can prevent pods of statefulsets from continually rolling.
```

**Comment by BenTheElder** [comment]
Verify is failing on a gofmt check, can you please run hack/update-gofmt.sh?

**Comment by BenTheElder** [comment]
one more format issue:
```
diff -u ./pkg/controller/history/controller_history_test.go.orig ./pkg/controller/history/controller_history_test.go
--- ./pkg/controller/history/controller_history_test.go.orig	2018-09-07 09:43:31.562334296 +0000
+++ ./pkg/controller/history/controller_history_test.go	2018-09-07 09:43:31.563334384 +0000
@@ -1545,7 +1545,7 @@
 		want      []string
 	}
 	testFn := func(test *testcase, t *testing.T) {
-		t.Run(test.name,func(t *testing.T) {
+		t.Run(test.name, func(t *testing.T) {
 			SortControllerRevisions(test.revisions)
 			for i := range test.revisions {
 				if test.revisions[i].Name != test.want[i] {

Run ./hack/update-gofmt.sh
```

**Comment by ryanmcnamara** [comment]
pull-kubernetes-verify — Job succeeded.
so I think format's good, not sure if the other failures were a flake or what

**Comment by BenTheElder** [comment]
Those do look like flakes, I've also mentioned this PR in #sig-apps on slack seeing if I can get someone appropriate to take a look. 👍

File: pkg/controller/history/controller_history.go
Comment (janetkuo) [inline pkg/controller/history/controller_history.go]: You don't need to sort it twice. In the `Less` function of `byRevision`, use a tie-breaker there when two revisions are equal. I'd prefer using something like `CreationTimestamp` as a tie-breaker.

Comment (janetkuo) [reply pkg/controller/history/controller_history.go]: We can use `Name` as another tie-breaker if both `Revision` and `CreationTimestamp` are equal.

**Comment by BenTheElder** [comment]
gentle poke @ryanmcnamara :-) https://github.com/kubernetes/kubernetes/pull/66882#discussion_r216516635

**Comment by BenTheElder** [comment]
Thanks for keeping after this! Poke @janetkuo :-)

**Comment by ryanmcnamara** [comment]
@janetkuo any update?

**Comment by ryanmcnamara** [comment]
Ping @BenTheElder @janetkuo

**Review by BenTheElder — APPROVED** [review]
/lgtm

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
@ryanmcnamara thank you. Would you squash commits?

**Comment by ryanmcnamara** [comment]
@janetkuo done!
Out of curiosity does the squash and merge github UI button not work? Or is there some additional process in merging prs for kubernetes?

**Comment by BenTheElder** [comment]
> Out of curiosity does the squash and merge github UI button not work? Or is there some additional process in merging prs for kubernetes?

We have a merge robot that ~ merges approved + lgtm'd PRs once they are passing tests against the target branch.

**Review by BenTheElder — APPROVED** [review]
/lgtm
sorry for the huge delay!

**Comment by BenTheElder** [comment]
Thanks for working on this! Merge details (`tide`) [here](https://prow.k8s.io/pr?query=is%3Apr+repo%3Akubernetes%2Fkubernetes+author%3Aryanmcnamara+head%3Arm%2Fstable-controllerrevisions-sort). This PR still needs approval from @janetkuo

**Comment by ryanmcnamara** [comment]
ping @janetkuo @BenTheElder

**Comment by BenTheElder** [comment]
/test pull-kubernetes-godeps
Will need Janet's approval still, or someone else in the OWNERS https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/history/OWNERS

**Review by janetkuo — APPROVED** [review]

**Comment by BenTheElder** [comment]
> godep: error downloading dep (bitbucket.org/ww/goautoneg): exit status 255

ah, we've had issues with hg + bitbucket, there's a fix being worked on. I don't think we have a tracking issue yet.

**Pull request merged**
